src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/AbstractApiAction.java

```diff
@@ -401,19 +401,23 @@ protected final RestChannelConsumer prepareRequest(RestRequest request, NodeClie
 				.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
 		final Object originalOrigin = threadPool.getThreadContext().getTransient(ConfigConstants.OPENDISTRO_SECURITY_ORIGIN);
 
-		return channel -> {
-
-			try (StoredContext ctx = threadPool.getThreadContext().stashContext()) {
-
+		return channel -> threadPool.generic().submit(() -> {
+			try (StoredContext ignore = threadPool.getThreadContext().stashContext()) {
 				threadPool.getThreadContext().putHeader(ConfigConstants.OPENDISTRO_SECURITY_CONF_REQUEST_HEADER, "true");
 				threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_USER, originalUser);
 				threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS, originalRemoteAddress);
 				threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_ORIGIN, originalOrigin);
 
 				handleApiRequest(channel, request, client);
-
+			} catch (Exception e) {
+				log.error("Error processing request {}", request, e);
+				try {
+					channel.sendResponse(new BytesRestResponse(channel, e));
```

Comment:
1. How was it taking care of sending responses before adding this catch(Ex) code ?
2. Will this send entire stack trace or just the message ? Is it possible that we might expose something important here ?

Reply:
It was relying on the 
```
        } catch (Exception e) {
            responseChannel.sendResponse(new BytesRestResponse(responseChannel, e));
        }
```
in `RestController.dispatchRequest()`.

Reply:
Since you are re-raising the Exception in here, wouldn't this cause RestController to also write the response in channel ? Do we need to do it twice ?

Reply:
The exception is raised in the `Runnable()` submitted to `generic()` `ExecutionService` for execution and is not part of the `RestController. dispatchRequest ()` thread stack (only `submit()` is part of that stack now).